### PR TITLE
Implement string list concatenation

### LINKED ISSUE #7727: [FEA] concatenate array of strings
**Is your feature request related to a problem? Please describe.**
We would like to support Spark's `concat_ws` function that can take any combination of strings or arrays of strings.

**Describe the solution you'd like**
cudf already offers a number of string concat APIs that can take a table of strings and concat them. What I would like is the ability to take a single string column that is an array of strings and concatenate them just like the table APIs do.  With that and the existing table APIs we should be able to build up `concat_ws`

**Describe alternatives you've considered**
There is no good alternative  The arrays could be variable length, so we cannot use any of the existing APIs that all assume a fixed number of inputs.

**Additional context**
Ideally we would want APIs that can either take a scalar string as the separator or a column_view of strings as the separator.  If we could only get one of them, then the column_view version would be better.

### PR DESCRIPTION
**Pull request by ttnghia**
Given a lists column of strings (each row is a list of strings), this PR facilitates the concatenation of strings within each list.

For example:

```
s = [ {'aa', 'bb', 'cc'}, null, {null, 'dd'}, {'ee', 'ff'} ]
r = strings::concatenate_list_elements(s, '+++') 
r is ['aa+++bb+++cc', null, null, 'ee+++ff']
```

This PR is similar to Spark's `concat_ws`, and closes #7727.

### CODE REVIEW

**Comment by ttnghia** [comment]
This PR is not ready to merge: I'm waiting for some pending refactoring work for the files `strings/combine.hpp|cu` to be merged first.

@davidwendt: Please mention this PR when you have your PR up, so I can keep track of your PR and will rebase after yours is merged.

File: cpp/include/cudf/strings/combine.hpp
Comment (jrhemstad) [inline cpp/include/cudf/strings/combine.hpp:210]: This isn't really concatenating rows. It's concatenating the elements within a list. I'd go for `concatenate_list_elements`.

Comment (ttnghia) [reply cpp/include/cudf/strings/combine.hpp:210]: I like that. And if you have even a better name (shorter) then it's perfect 😄

File: cpp/src/strings/combine.cu
Comment (jrhemstad) [inline cpp/src/strings/combine.cu:578]: Don't do this. This is a memcpy outside of stream order.

**Review by davidwendt — CHANGES_REQUESTED** [review]
The changes to combine/concatenate.cu are not necessary for this PR.
Can you change it back?

**Comment by ttnghia** [comment]
> The changes to combine/concatenate.cu are not necessary for this PR.
> Can you change it back?

No problem. I just wanted to make it better.

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (davidwendt) [inline cpp/src/strings/combine/concatenate_list_elements.cu:161]: Using negative values here to identify nulls seems hacky since the offset value will need to be converted back to zero in the end.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/strings/utilities.cuh
Comment (davidwendt) [inline cpp/src/strings/utilities.cuh:162]: This whole utility is predicated on using -1 in the offsets/sizes to identify null entries and then changing it back to 0 during the scan. This seems wrong to me.

Comment (ttnghia) [reply cpp/src/strings/utilities.cuh:162]: The basic idea of using negative size is to identify null entries. Given that the column contains lists of strings, each list may hold a very large number of strings. Thus, both size computation and validity check can be very expensive. If we combine them, both expensive operations have to be computed separately. On the other hand, using a negative value to identify null entry then clamp it to zero to compute offsets seem to be perfect. We don't have any issue with negative value here.

Comment (davidwendt) [reply cpp/src/strings/utilities.cuh:162]: How about combining setting the sizes and the nullmask in a custom kernel instead? 
Then the valid_if launch would not be needed.

Comment (ttnghia) [reply cpp/src/strings/utilities.cuh:162]: That's possible, but the kernel to set bitmask would be more complex thus I'm a bit reluctant to do that. I'm thinking of a cleaner way to use an additional array of `int8_t` type to store the validity values (`0/1`), then just call `valid_if` on that array. How do you think about this solution?

**Comment by ttnghia** [comment]
@davidwendt the file `tests/combine_tests.cpp` has grown up significantly (~1000 LOC). I'm thinking of splitting it to 

```
tests/combine/concatenate_tests.cpp
tests/combine/concatenate_list_elements_tests.cpp
tests/combine/join_strings_tests.cpp
```

Do you agree with this split?

File: cpp/src/strings/utilities.cuh
Comment (hyperbolic2346) [inline cpp/src/strings/utilities.cuh:176]:
```suggestion
  // If all the strings are empty or null, `bytes` will be `0` thus the `d_chars` pointer will have
```

File: cpp/src/strings/utilities.cuh
Comment (hyperbolic2346) [inline cpp/src/strings/utilities.cuh:181]: This seems very suspect. It could be true today, but this is the type of thing that can change and present issues.

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]

**Comment by davidwendt** [comment]
> @davidwendt the file `tests/combine_tests.cpp` has grown up significantly (~1000 LOC). I'm thinking of splitting it to
> 
> ```
> tests/combine/concatenate_tests.cpp
> tests/combine/concatenate_list_elements_tests.cpp
> tests/combine/join_strings_tests.cpp
> ```
> 
> Do you agree with this split?

Yes.

File: cpp/src/strings/utilities.cuh
Comment (davidwendt) [reply cpp/src/strings/utilities.cuh:181]: It seems like we want  `for_each_fn` will do nothing if `bytes==0` here?
If so why not just do this instead:
```
if( bytes > 0 ) { 
  for_each_fn(size_and_exec_fn);
}
```

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (ttnghia) [reply cpp/src/strings/combine/concatenate_list_elements.cu:158]: Then we will have duplicate code with nearly the same functors. The overhead of executing the same functor for different tasks is very small.

File: cpp/tests/strings/combine/join_strings_tests.cpp
Comment (ttnghia) [inline cpp/tests/strings/combine/join_strings_tests.cpp:1]: The code in this file is not part of this PR. I just extract one large cpp test file into multiple files.

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (hyperbolic2346) [reply cpp/src/strings/combine/concatenate_list_elements.cu]: I'm not sure I follow. I'm asking to remove the duplicated code.

File: cpp/src/strings/utilities.cuh
Comment (ttnghia) [reply cpp/src/strings/utilities.cuh]: Oh yeah, why not!!!

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (ttnghia) [reply cpp/src/strings/combine/concatenate_list_elements.cu]: Ah I'm sorry, I misunderstood that. Good point, I'll look into it. Although the 2 functors can't call each other directly, but I can refactor to remove some duplicate code.

File: cpp/src/strings/utilities.cuh
Comment (ttnghia) [reply cpp/src/strings/utilities.cuh]: Done.

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (ttnghia) [reply cpp/src/strings/combine/concatenate_list_elements.cu]: Done. Please have a look. No more duplicate code now. Thanks for the idea.

**Comment by harrism** [comment]
> Given a lists column of strings (each row is a list of strings), this PR facilitates the concatenation of strings within each list.
> 
> For example:
> 
> ```
> s = [ {'aa', 'bb', 'cc'}, null, {null, 'dd'}, {'ee', 'ff'} ]
> concatenate_rows(s, "+++") = ['aa+++bb+++cc', null, null, 'ee+++ff']
> ```
> 
> This PR also partially supports Spark's `concat_ws` (closes #7727).

I think the name `concatenate_rows` is wrong for this API behavior. The behavior demonstrated above is concatenating the strings within each row into a single string. It is not concatenating multiple rows of a single column into one row, or concatenating corresponding rows of multiple columns into one row of an output column. Either of those might be called `concatenate_rows`. 

I would call this `concatenate_strings` or `concatenate_strings_list` or something.

**Comment by ttnghia** [comment]
> > Given a lists column of strings (each row is a list of strings), this PR facilitates the concatenation of strings within each list.
> > For example:
> > ```
> > s = [ {'aa', 'bb', 'cc'}, null, {null, 'dd'}, {'ee', 'ff'} ]
> > concatenate_rows(s, "+++") = ['aa+++bb+++cc', null, null, 'ee+++ff']
> > ```
> > 
> > 
> > This PR also partially supports Spark's `concat_ws` (closes #7727).
> 
> I think the name `concatenate_rows` is wrong for this API behavior. The behavior demonstrated above is concatenating the strings within each row into a single string. It is not concatenating multiple rows of a single column into one row, or concatenating corresponding rows of multiple columns into one row of an output column. Either of those might be called `concatenate_rows`.
> 
> I would call this `concatenate_strings` or `concatenate_strings_list` or something.

Thanks Mark. The name `concatenate_strings_list` sounds good to me. Jake has also suggested the name `concatenate_list_elements` and I have adopted it (sorry that I didn't update the API name in the example). 

So, @jrhemstad now we have one more name suggestion. For me, `concatenate_strings_list` would sound better. How do you think?

**Comment by harrism** [comment]
`concatenate_list_elements` works for me. Will there ever be a type stored in a list other than strings that can be concatenated?

**Comment by ttnghia** [comment]
> `concatenate_list_elements` works for me and I like it better than my suggestion. However, will there ever be a type stored in a list other than strings that can be concatenated?

Here the API is part of the `strings` namespace and it only works for strings type.

Alright, it seems that `concatenate_list_elements` is a good fit, because in the future we may have other `concatenate_list_elements` APIs in other namespaces, such as an API for concatenating list elements within a list (for example, something that is similar to list unravel here https://nvidia.slack.com/archives/CB9GTQZAQ/p1618926053144000).

File: cpp/tests/strings/combine/concatenate_list_elements_tests.cpp
Comment (davidwendt) [inline cpp/tests/strings/combine/concatenate_list_elements_tests.cpp:36]:
```suggestion
constexpr bool print_all{false};
```

File: cpp/src/strings/utilities.cuh
Comment (davidwendt) [inline cpp/src/strings/utilities.cuh:146]: Considering I foresee this is only to be used by the `concatenate_list_elements` I wonder if it should not be added to this include file since this file is included by many, many other source files in libcudf increasing compile time.

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (davidwendt) [inline cpp/src/strings/combine/concatenate_list_elements.cu:2]:
```suggestion
 * Copyright (c) 2021, NVIDIA CORPORATION.
```

**Review by davidwendt — CHANGES_REQUESTED** [review]
This is looking good.

File: cpp/src/strings/utilities.cuh
Comment (ttnghia) [reply cpp/src/strings/utilities.cuh:146]: I thought about that too, but I was not sure whether we will need this utility function in the future for other APIs. Currently, I'm working on row-wise lists concatenation and lists column interleaving, which also include operations for string column. They are about to be completed soon and I'll let you know if this utility is needed there.

File: cpp/src/strings/combine/concatenate_list_elements.cu
Comment (ttnghia) [reply cpp/src/strings/combine/concatenate_list_elements.cu:102]: Yes.

File: cpp/src/strings/utilities.cuh
Comment (ttnghia) [reply cpp/src/strings/utilities.cuh:146]: @davidwendt I think we should leave the utility function in that header, as it would be many potential usages. In addition, I would like to rewrite the implementation in `combine.cu` (in a separate PR) to use it---that would make the code a lot shorter.

Comment (davidwendt) [reply cpp/src/strings/utilities.cuh:146]: I'm ok with leaving it as utility if you think it will be valuable to your other work. 
If that work is outside of `cpp/src/strings/` then we may want to move this utility to `cpp/include/cudf/strings/detail/utilities.cuh` instead.
I'm not ok with rewriting combine/concatenate.cu to use it.

**Review by davidwendt — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**
